services/requester/pool.go

```diff
@@ -102,6 +154,224 @@ func (t *TxPool) Send(
+
+	singleTxPool := NewSingleTxPool(
+		client,
+		transactionsPublisher,
+		logger,
+		config,
+		collector,
+		keystore,
+	)
+	batchPool := &BatchTxPool{
+		SingleTxPool: singleTxPool,
+		pooledTxs:    make(map[gethCommon.Address][]pooledEvmTx),
+		txMux:        sync.Mutex{},
+	}
+
+	go batchPool.processPooledTransactions()
+
+	return batchPool
+}
+
+// Add adds the EVM transaction to the tx pool, grouped with the rest of the
+// transactions from the same EOA signer.
+// After the configured `TxBatchInterval`, the collected transations
+// are batched and sent to the Flow network using `EVM.batchRun`, for execution.
+func (t *BatchTxPool) Add(
+	ctx context.Context,
+	tx *gethTypes.Transaction,
+) error {
+	// tx sending should be blocking, so we don't have races when
+	// pooled transactions are being processed in the background.
+	t.txMux.Lock()
+	defer t.txMux.Unlock()
+
+	t.txPublisher.Publish(tx) // publish pending transaction event
+
+	from, err := gethTypes.Sender(gethTypes.LatestSignerForChainID(tx.ChainId()), tx)
+	if err != nil {
+		return fmt.Errorf("failed to derive the sender: %w", err)
+	}
+	txData, err := tx.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	hexEncodedTx, err := cadence.NewString(hex.EncodeToString(txData))
+	if err != nil {
+		return err
+	}
+
+	userTx := pooledEvmTx{txPayload: hexEncodedTx, nonce: tx.Nonce()}
+	t.pooledTxs[from] = append(t.pooledTxs[from], userTx)
+
+	return nil
+}
+
+func (t *BatchTxPool) processPooledTransactions() {
+	for range time.Tick(t.config.TxBatchInterval) {
+		// Take a snapshot here to allow `Add()` continue accepted
+		// incoming EVM transactions, without blocking until the
+		// batch transactions are submitted.
+		snapshot := func() map[gethCommon.Address][]pooledEvmTx {
+			t.txMux.Lock()
+			defer t.txMux.Unlock()
+			copy := t.pooledTxs
+			t.pooledTxs = make(map[gethCommon.Address][]pooledEvmTx)
+			return copy
+		}()
+
+		for address, pooledTxs := range snapshot {
+			if err := t.batchSubmitTransactions(pooledTxs); err != nil {
+				t.logger.Error().Err(err).Msgf(
+					"failed to send Flow transaction from BatchPool for EOA: %s",
+					address.Hex(),
+				)
+				continue
+			}
+		}
+	}
+}
+
+func (t *BatchTxPool) batchSubmitTransactions(
+	pooledTxs []pooledEvmTx,
+) error {
+	// Sort the transactions based on their nonce, to make sure
+	// that no re-ordering has happened due to races etc.
+	sort.Slice(pooledTxs, func(i, j int) bool {
+		return pooledTxs[i].nonce < pooledTxs[j].nonce
+	})
+
+	hexEncodedTxs := make([]cadence.Value, len(pooledTxs))
+	for i, txPayload := range pooledTxs {
+		hexEncodedTxs[i] = txPayload.txPayload
+	}
+
+	coinbaseAddress, err := cadence.NewString(t.config.Coinbase.Hex())
+	if err != nil {
+		return err
+	}
+
+	script := replaceAddresses(batchRunTxScript, t.config.FlowNetworkID)
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	flowTx, err := t.buildTransaction(
```

Comment:
I think we should refactor to reuse the block and account across all cadence tx sent in the batch.

this is going to do lookups for the latest block and account for each EOA. Calls to get account take anywhere from 25-500ms on the public endpoints. nodes with a medium/high number of keys probably take somewhere in the middle. If the node is sending data for many EOAs, this could easily take longer than 1 second to send all tx.

We can definitely reuse the block. it looks like we can generally reuse the account. The only data that really changes is the key sequence numbers, and we already have the limitation that there must be enough keys to handle 10-15 seconds worth of transactions. We will actually use fewer keys with batching.